src/main/java/org/elasticsearch/index/analysis/TurkishStemmer.java

```diff
@@ -184,6 +184,33 @@ public String stem(char s[], int len) {
     wordsToStem = new HashSet<String>(stems);
     wordsToStem.add(originalWord);
 
+    if (wordsToStem.contains(new String(s, 0, len)) && wordsToStem.size() < 2
+            && d < 1) {
+      // If none of the stemming rules matches
+        char lastLetter = s[len - 1];
+        if (lastLetter == 'u' || lastLetter == 'ü' || lastLetter == 'i' ||
+                lastLetter == 'ı') {
+          // if it's a u replace it with ü and vice versa
+          if (lastLetter == 'u') {
+            s[len - 1] = 'ü';
+          } else if (lastLetter == 'ü') {
+            s[len - 1] = 'u';
+            // if it's a i replace it with ı and vice versa
+          } else if (lastLetter == 'i') {
+            s[len - 1] = 'ı';
+          } else if (lastLetter == 'ı') {
+            s[len - 1] = 'i';
+          }
+          // and try stemming again - hopefully it will find a match this time
+          // recursive call - beware!
+          // TODO: add an argument to restrict number of potential calls in order
+          // to avoid worst case scenario -> stack overflow
+          // TODO: check if there are other cases we need to consider e.g.
+          // vowel harmony
+          return stem(s, len, ++d);
```

Comment:
Is this not possible to cause a `StackOverflowError` ?
There seems to be no recursion breaker